src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/DocumentPresentation/TextDocumentUriPresentationEndpoint.cs

```diff
@@ -0,0 +1,225 @@
+            _razorDocumentMappingService = razorDocumentMappingService;
+            _razorComponentSearchEngine = razorComponentSearchEngine;
+            _logger = loggerFactory.CreateLogger<TextDocumentUriPresentationEndpoint>();
+        }
+
+        public RegistrationExtensionResult GetRegistration()
+        {
+            const string AssociatedServerCapability = "_vs_uriPresentationProvider";
+
+            return new RegistrationExtensionResult(AssociatedServerCapability, true);
+        }
+
+        public async Task<WorkspaceEdit?> Handle(UriPresentationParams request, CancellationToken cancellationToken)
+        {
+            var documentSnapshot = await _projectSnapshotManagerDispatcher.RunOnDispatcherThreadAsync(() =>
+            {
+                _documentResolver.TryResolveDocument(request.TextDocument.Uri.GetAbsoluteOrUNCPath(), out var documentSnapshot);
+
+                return documentSnapshot;
+            }, cancellationToken).ConfigureAwait(false);
+
+            if (documentSnapshot is null)
+            {
+                _logger.LogWarning($"Failed to find document {request.TextDocument.Uri}.");
+                return null;
+            }
+
+            cancellationToken.ThrowIfCancellationRequested();
+
+            var codeDocument = await documentSnapshot.GetGeneratedOutputAsync().ConfigureAwait(false);
+            if (codeDocument.IsUnsupported())
+            {
+                _logger.LogWarning($"Failed to retrieve generated output for document {request.TextDocument.Uri}.");
+                return null;
+            }
+
+            var sourceText = await documentSnapshot.GetTextAsync().ConfigureAwait(false);
+            if (request.Range?.Start.TryGetAbsoluteIndex(sourceText, _logger, out var hostDocumentIndex) != true)
+            {
+                return null;
+            }
+
+            var languageKind = _razorDocumentMappingService.GetLanguageKind(codeDocument, hostDocumentIndex);
+            if (languageKind is not RazorLanguageKind.Html)
+            {
+                _logger.LogInformation($"Unsupported language {languageKind:G}.");
+                return null;
+            }
+
+            if (request.Uris is null || request.Uris.Length == 0)
+            {
+                _logger.LogInformation($"Didn't get any Uris?");
+                return null;
+            }
+
+            // We only want to handle requests for a single .razor file, but when there are files nested under a .razor
+            // file (for example, Goo.razor.css, Goo.razor.cs etc.) then we'll get all of those files as well, when the user
+            // thinks they're just dragging the parent one, so we have to be a little bit clever with the filter here
+            var razorFileUri = request.Uris.Last();
+            var fileName = Path.GetFileName(razorFileUri.GetAbsoluteOrUNCPath());
+            if (!Path.GetExtension(fileName).Equals(".razor", StringComparison.OrdinalIgnoreCase))
```

Comment:
Doing an EndsWith might be better than extracting the extension. Also can pass in FilePathComparison.Instance